### ISSUE TITLE
Kiali dashboard freezes when checking traces information

### ISSUE THREAD

**Issue by Myeongseok-Kim** [kiali/kiali#5584]
After enabling Jaeger, I confirmed that I can check traces information in kiali.
![image](https://user-images.githubusercontent.com/54100684/197927234-091e8510-70eb-4566-a97e-7356f80e782b.png)

But when I try to check traces information
Something seems to be happening in the Kiali dashboard.
(Loading animation plays)
![image](https://user-images.githubusercontent.com/54100684/197927173-54b6b428-8ef8-485a-8ef8-df42557990b3.png)
This puts a load on my prometheus server, and eventually the Prometheus server went down due to OOM.
This issue is still occurring.
Does it only happen in my environment?

**Version information**
istio: 1.15.2
kiali: 1.58.0
jaeger: 1.38.0
prometheus: 2.37.0

**Comment by jshaughn** [comment]
Hi @Myeongseok-Kim,  a few questions:
- Is the problem repeatable?
- When you say, `But when I try to check traces information` do you mean that you clicked on one of the traces in the chart? Or was it some other action, like moving the mouse over the chart to look at tooltip information? 
- Did you click the "Distributed Tracing" option in the main menu, or are you on the "Traces" tab of a workload detail? 
- Does the rest of Kiali function well, for example, can you generate a graph with no problem?
- Is there any reason to think the issue may be with you prometheus pod resource limits or connectivity?

**Comment by jshaughn** [comment]
@Myeongseok-Kim I think maybe I see a problem but let me ask, after the traces chart is displayed, did you at any point drag the pointer across a large number of trace nodes, or look at many of the tooltips?  Also, did this problem happen quickly or had you been interacting with the chart over a longer period of time?

**Comment by Myeongseok-Kim** [comment]
> Hi @Myeongseok-Kim, a few questions:
> 
> * Is the problem repeatable?
> * When you say, `But when I try to check traces information` do you mean that you clicked on one of the traces in the chart? Or was it some other action, like moving the mouse over the chart to look at tooltip information?
> * Did you click the "Distributed Tracing" option in the main menu, or are you on the "Traces" tab of a workload detail?
> * Does the rest of Kiali function well, for example, can you generate a graph with no problem?
> * Is there any reason to think the issue may be with you prometheus pod resource limits or connectivity?


- It is possible to reproduce the problem.
- Occurs when the mouse hovers over a point on the tracking chart.
(This simply happens without any other action.)
- Occurs in the "traces" tab of the workload details. There is no problem loading the chart, it happens when you mouse over it.
- Before the problem occurs, kiali works fine. However, when this problem occurs, the graph is not even drawn.
- My Prometheus is now using 2GB of memory, and can use up to 6GB.
(There is no limit, but there is a limit to the node's memory.)
-  Memory usage goes up steeply when I do problematic behavior. If I keep the kiali browser, Prometheus will OOM, and when I close the Kiali browser, the memory of Prometheus goes back to normal.

**Comment by Myeongseok-Kim** [comment]
> @Myeongseok-Kim I think maybe I see a problem but let me ask, after the traces chart is displayed, did you at any point drag the pointer across a large number of trace nodes, or look at many of the tooltips? Also, did this problem happen quickly or had you been interacting with the chart over a longer period of time?

- The period of time is irrelevant.
- Kiali > Workloads > Select Workload > Traces tab > Mouse over chart points
- This is the pattern of occurrence of this problem.

**Comment by jshaughn** [comment]
The problem has to do with the fact that a backend call is sometimes made to help generate the heatmap in the tooltip.  This was not really a problem in the past, but I think with more recent updates to Patternfly (the ui component library) and/or Victory charts (the charting library), that the user no longer has to hover over the chart point to activate the tooltip, instead, whenever the mouse is anywhere in the chart the library "snaps" to the nearest chart point and tries to generate the tooltip.  This is bad because if you drag across the chart it is snapping to many points as the mouseover location moves.  This can then cause many concurrent calls to the backend for metric data to build the tooltip.   We need to stop that instant snap behavior.  Either we need to configure a delay, force an actual hover (no snap), or throttle requests in some other way.

**Comment by jshaughn** [comment]
Fixed by #5600

**Comment by Myeongseok-Kim** [comment]
hello.
I found a fix for that issue in the V1.59.0 release notes.
Unfortunately, my problem seems to have other causes.
The problem arises even if I don't traverse multiple points, and carefully hover over one point.

![image](https://user-images.githubusercontent.com/54100684/200449071-225d8b44-30d3-4056-80c8-7fd0ecf6b35e.png)

This is less of a problem, but sometimes it happens when there are only 2 points.
![image](https://user-images.githubusercontent.com/54100684/200449661-e5f47be7-cccb-4ec9-a18c-a899cdfb0c1a.png)

**Comment by jshaughn** [comment]
Please first verify that you are truly running v 1.59 by visiting the About Box in the Kiali masthead.  Assuming the version is correct it's not obvious what this second issue could be.  Is Kiali freezing AND prometheus still suffering OOM?  Or just one of these problems is now occurring?

**Comment by Myeongseok-Kim** [comment]
I have applied v1.59.0 version.
![image](https://user-images.githubusercontent.com/54100684/200485508-316fe45c-8480-4fca-bbdc-aab457963be9.png)

Hovering over the point triggers the kiali action animation, which seems to last until I refresh the browser.
The span detail tab is also shown at this time. Other menus load slower than usual.
The resource usage of Prometheus increases rapidly, and I close the Kiali browser to prevent OOM. OOM will happen if I don't close the browser.

**Comment by jshaughn** [comment]
I'm not sure what could cause this problem and I can not reproduce the behavior.   It seems like maybe you are getting slow response times from Prometheus and it affects things in an unexpected way.
Do you see any console errors when this happens?
Do you see the "heatmap" tooltips when mousing over the chart and hovering on a trace point?
Does the problem happen only after clicking a trace point to see the Trace/Span detail below the chart?
What is your browser?  And please make sure your browser cache has been cleared after the update.

**Comment by jshaughn** [comment]
I don't understand why you have concurrent requests for `metrics` , the change in v1.59 was supposed to fix that problem.  But the real problem is the huge latency for those calls.  They should return in a few ms, not 15+ seconds.
- Does the rest of Kiali operate normally?
- How long does it take to generate a graph?
   - without Edge Labels
   - with Response Time Edge Labels
- How about to generate the charts for the `Outbound Metrics` or `Inbound  Metrics` tab (assuming there is traffic)?

It's possible your prometheus instance is not performing well.  We need to understand if it is your environment or if there is really an issue with this specific use case.  It's very possible that you are storing too many metrics for your infrastructure and may need to tune prometheus.

**Comment by Myeongseok-Kim** [comment]
- Does the rest of Kiali operate normally?
=> Except for the Traces tab, everything else seems to be working fine.

- How long does it take to generate a graph?
=> There is no long latency in graph generation.
  - without Edge Labels
![image](https://user-images.githubusercontent.com/54100684/200968498-41bf94b0-574e-4e39-ba7d-6716187609b8.png)

  - with Response Time Edge Labels
![image](https://user-images.githubusercontent.com/54100684/200968595-87b4c28b-f073-4bba-a48e-690ed3a0def4.png)

- How about to generate the charts for the Outbound Metrics or Inbound Metrics tab (assuming there is traffic)?
 => There is no problem with generating metric charts.
![image](https://user-images.githubusercontent.com/54100684/200968716-79d355d3-259e-4fff-a088-f4730dda88b4.png)

**Comment by jshaughn** [comment]
Thanks for answering all of those questions.  I'm glad the rest of Kiali is responding well; it would indicate that probably you don't have an issue with prometheus itself, or the amount of metric storage.  My next thought was that your traces may have a very large number of spans, but looking back it seems there are only a couple of spans in the screenshots above.
- Is the trace chart itself appearing reasonably fast?
- Do you have any problem with a trace overlay in the graph?
  - On the graph click on a workload.  Select the `Traces` tab on the side panel.  Select a trace and you should see an overlay in the graph for the trace spans.

I will continue to look at the tooltip code for clues.

**Comment by jmazzitelli** [comment]
Since the problem is with traces, wouldn't the potential bottleneck be Jaeger and not Prometheus? How is Jaeger performance?

**Comment by Myeongseok-Kim** [comment]
> Thanks for answering all of those questions. I'm glad the rest of Kiali is responding well; it would indicate that probably you don't have an issue with prometheus itself, or the amount of metric storage. My next thought was that your traces may have a very large number of spans, but looking back it seems there are only a couple of spans in the screenshots above.
> 
> * Is the trace chart itself appearing reasonably fast?
> * Do you have any problem with a trace overlay in the graph?
>   
>   * On the graph click on a workload.  Select the `Traces` tab on the side panel.  Select a trace and you should see an overlay in the graph for the trace spans.
> 
> I will continue to look at the tooltip code for clues.


There is no delay in generating the trace chart itself.
In the Graph tab, the traces in the workload sidepanel work fine.
![image](https://user-images.githubusercontent.com/54100684/201240082-71a331f0-e9fd-46a7-86b2-43710516fb20.png)

**Comment by Myeongseok-Kim** [comment]
> Since the problem is with traces, wouldn't the potential bottleneck be Jaeger and not Prometheus? How is Jaeger performance?

I also don't understand why the load on Prometheus is so high in the traces.
Jaeger UI also works fine.
My service doesn't have a lot of traffic, Jaeger is only collecting 5% sampling.

**Comment by jshaughn** [comment]
> There is no delay in generating the trace chart itself.  In the Graph tab, the traces in the workload sidepanel work fine.

@Myeongseok-Kim I am still investigating.  In the graph side-panel, can you try clicking one of the traces in the list.  I don't think there is a problem but make sure the trace overlay appears on the graph, and the trace detail shows up in the side panel.

**Comment by jshaughn** [comment]
@Myeongseok-Kim In your screenshot above I see several requests in a "Pending" state on the browser.  That's fairly strange.  I wonder if something is blocking this particular request, on the client side.  Can you try testing this in an incognito window on the browser and see if the behavior is different?

**Comment by Myeongseok-Kim** [comment]
> > There is no delay in generating the trace chart itself.  In the Graph tab, the traces in the workload sidepanel work fine.
> 
> @Myeongseok-Kim I am still investigating. In the graph side-panel, can you try clicking one of the traces in the list. I don't think there is a problem but make sure the trace overlay appears on the graph, and the trace detail shows up in the side panel.

This is the result of clicking on one of the trace lists in the side panel.
![image](https://user-images.githubusercontent.com/54100684/201819817-12e955e0-b9db-4bda-befb-fbbd4955f5bb.png)

**Comment by Myeongseok-Kim** [comment]
> @Myeongseok-Kim In your screenshot above I see several requests in a "Pending" state on the browser. That's fairly strange. I wonder if something is blocking this particular request, on the client side. Can you try testing this in an incognito window on the browser and see if the behavior is different?
> 
> One more request, if the behavior is still poor, for any long-running `metrics` network request (for example, in your screenshot above:
> 
> ![image](https://user-images.githubusercontent.com/2104052/201685002-58bc757a-329a-4447-a913-610ca7db7b18.png)
> 
> click on the one of those `metrics` rows, click on the `Payload` tab, expand the `queries` and capture that screen shot, or minimally the number of `query` elements. Thanks,


The same goes for incognito windows.
It looks like 2 identical requests are happening.

![image](https://user-images.githubusercontent.com/54100684/201820420-787537d7-f2e3-44d1-937e-8f700bb4ba09.png)
![image](https://user-images.githubusercontent.com/54100684/201820449-b6e9a45c-daa1-439f-b095-364a882a6a44.png)

**Comment by jshaughn** [comment]
@Myeongseok-Kim Thanks for the information, I will continue to investigate.

**Comment by jshaughn** [comment]
Because this issue is closed, and because this looks like a separate problem, I have opened a new issue #5634.  Discussion should continue in that issue.
